Pull request from empty_branch to master

Swapnil-ingle opened the pull request.

Swapnil-ingle committed:
SI: Added the code for creating Collection Protocols according to the CARS CSV file.

Swapnil-ingle committed:
SI: Fixed msk-cars.xml intendation, Removed extra spaces, changed variable cpid --> cpID, the importCp() method now throws Exception.

Swapnil-ingle committed:
SI: Removed the 'msk' prefix from all classes, Changed CARS --> Cars, Changed variabled names in CarsDetail according to CARS domain, setCpSites() method code made simplified and coherent, the directory structure for msk-cars.xml changed to 'src/main/resources/com/krishagni/openspecimen/msk2/schema/', the i18n keys named changed to msk_cars_irb_number, msk_cars_facility and msk_cars_pi, the CpImport() method now returns resp.payload()

Swapnil-ingle committed:
SI: Moved the plugin initialiser bean from pluginApiContext.xml to pluginContext.xml

Swapnil-ingle committed:
SI: Deleted pluginApiContext.xml file

Swapnil-ingle committed:
SI: Implemented the in-memory map(cpMap, eventsMap, srMap) pseudo code
1. Created the three maps as per pseudo code.
2. Populated the maps as per pseudo code. (Currently not doing add/update anything to DB, just populating the maps created in #1)

Swapnil-ingle committed:
SI: Added the functionality for createVisit(), createSr()
1. Added the implementation of createVisit() method
2. Added the implementation of createSr() method.
3. Did some minor intendation changes.
4. Removed the unneccessary comments.

Swapnil-ingle committed:
SI: Added the functionality for updating Specimen Requirement.
1. Implemented the method updateSr()
2. Made changes to getSrFromDb() method
3. Minor indentation change

Swapnil-ingle committed:
SI: Added low level redirections for the method checkCP, checkSr, checkEvent and removed the hard-coded string.
1. Moved the code to lower level for:
  1. checkCp()
  2. checkEvent()
  3. checkSr()
2. Removed the hard-coded Strings and made them private final static instead

vinayakapawar commented on src/main/java/com/krishagni/openspecimen/msk2/importer/CarsImporter.java:
```diff
@@ -0,0 +1,278 @@
+	
+	@PlusTransactional
+	private ResponseEvent<Object> importRecord(ImportObjectDetail<CarsDetail> detail) throws Exception {
+		CollectionProtocolEventDetail event = toEvent(detail.getObject());
```
I would do this as rather than overloading a single function with multiple responsibilities

CollectionProtocolDetail cp = toCp(detail.getObject());
CollectionProtocolEventDetail event = toEvent(detail.getObject());
SpecimenRequirementDetail  sr = toSr(detail.getObject());

vinayakapawar commented on src/main/java/com/krishagni/openspecimen/msk2/importer/CarsImporter.java:
```diff
@@ -0,0 +1,278 @@
+
+	private CollectionProtocolEventDetail toEvent(CarsDetail detail) {
+		CollectionProtocolEventDetail eventDetail = new CollectionProtocolEventDetail();
+		SpecimenRequirementDetail srDetail = new SpecimenRequirementDetail();
```
Variables should be initialised at the point of their first usage. This is not C where you initialise all the variable at the beginning of the block.

vinayakapawar commented on src/main/java/com/krishagni/openspecimen/msk2/importer/CarsImporter.java:
```diff
@@ -0,0 +1,278 @@
+		srDetail.setName(detail.getProcedureName());
+		srDetail.setInitialQty(new BigDecimal(0));
+		srDetail.setCollectionContainer(detail.getCollectionContainer());
+		srDetail.setSpecimenClass(getSpecimenClass(detail.getSpecimenType()));
```
1. The order of fields initialisation will be from higher order to lower order.
2. The higher order fields are those that help in recognising a record from others.
3. In this case, it would be - cp short title, event label, name, type.
4. The rest can appear in whichever order you like

vinayakapawar commented on src/main/java/com/krishagni/openspecimen/msk2/importer/CarsImporter.java:
```diff
@@ -0,0 +1,278 @@
+		eventDetail.setEventLabel(detail.getCycleName() + detail.getTimepointName());
+		
+		srDetail.setCpShortTitle(detail.getIrbNumber());
+		srDetail.setType(detail.getSpecimenType());
```
1. The order of fields initialisation will be from higher order to lower order.
2. The higher order fields are those that help in recognising a record from others.
3. In this case, it would be - cp short title, event label, name, type.
4. The rest can appear in whichever order you like

vinayakapawar commented on src/main/java/com/krishagni/openspecimen/msk2/importer/CarsImporter.java:
```diff
@@ -0,0 +1,278 @@
+		srDetail.setName(detail.getProcedureName());
+		srDetail.setInitialQty(new BigDecimal(0));
+		srDetail.setCollectionContainer(detail.getCollectionContainer());
+		srDetail.setSpecimenClass(getSpecimenClass(detail.getSpecimenType()));
```
I don't think you need to explicitly initialise the class. The specimen API will take care of finding the class from the type.

Swapnil-ingle replied:
@vinayakapawar 
I tried adding an SR from the Postman via Specimen Rest-API POST method:
1. Gave Specimen class as "Not Specified": Error(Specimen class is invalid)
2. Did not give an input for Specimen class: Error(Specimen class is required)

The Specimen API internally calls POST method in SpecimenRequirementsController class.
```java
cpSvc.addSpecimenRequirement(request(requirement))
```
While inspecting how UI does not ask for Specimen Class, I found that the UI calls permissible-values/ API and fills the request header with the all information needed, including Specimen class, and then sends this information to the backend.

If cpSvc.addSpecimenRequirement() did not require Specimen class, one of the above #1 and #2 should've worked, right?

Is this information correct? What Specimen API were you referring to that'll handle adding Specimen class by itself?

vinayakapawar replied:
OK. I see. I got confused with the specimen API. The specimen API doesn't have class as the mandatory attribute. Fine go ahead with lookup and populate the correct class.

vinayakapawar commented on src/main/java/com/krishagni/openspecimen/msk2/importer/CarsImporter.java:
```diff
@@ -0,0 +1,278 @@
+		srDetail.setSpecimenClass(getSpecimenClass(detail.getSpecimenType()));
+		
+		eventDetail.setSpecimenRequirements(Arrays.asList(srDetail));
+		
```
This will go away when you implement single responsibility methods.

vinayakapawar commented on src/main/java/com/krishagni/openspecimen/msk2/importer/CarsImporter.java:
```diff
@@ -0,0 +1,278 @@
+	}
+
+	private String getSpecimenClass(String specimenType) {
+		ListPvCriteria crit = new ListPvCriteria()
```
This method is not needed. See my above comment.

vinayakapawar commented on src/main/java/com/krishagni/openspecimen/msk2/importer/CarsImporter.java:
```diff
@@ -0,0 +1,278 @@
+	//
+	/////////////////////////
+	
+	private CollectionProtocolDetail checkCp(CollectionProtocolEventDetail event, CarsDetail object) throws Exception {
```
1. The input to this method should be just CollectionProtocolEventDetail
2. The method name should be getCp(CollectionProtocolDetail cp)

vinayakapawar commented on src/main/java/com/krishagni/openspecimen/msk2/importer/CarsImporter.java:
```diff
@@ -0,0 +1,278 @@
+		if (cpDetail == null) {
+			cpDetail = getCpFromDB(event.getCpShortTitle());
+			if (cpDetail == null) {
+		        	cpDetail = createCp(object);
```
It is not a good idea to have nested conditionals. It breaks the program linearity.

Instead, do this:

```
public CollectionProtocolDetail getCp(CollectionProtocolDetail input) {
  return cpMap.computeIfAbsent(input.getShortTitle(), () -> createIfAbsent(input));
}

private CollectionProtocolDetail createIfAbsent(CollectionProtocolDetail input) {
  CollectionProtocolDetail dbCp = getCpFromDb(input.getShortTitle());
  return dbCp != null ? dbCp : createCp(input);
}
```

Swapnil-ingle replied:
@vinayakapawar 
The createIfAbsent() method here throws an Exception (as getCpFromDb() method throws an Exception).
When I'm using cpMap.computeIfAbsent() I have to try-catch and handle the exception right there in getCp() method but we have to throw the exception to the upper method, as that one is @PlusTransactional. How do I propagate the Exception to upper method.
I searched for how to throw Exceptions from computeIfAbsent() method, throwing Exception from Lambda expressions, etc. There seems no way around it. 

I'm stumped, Can you please help?

```java
private CollectionProtocolDetail getCp(CollectionProtocolDetail input) throws Exception {
		return cpMap.computeIfAbsent(input.getShortTitle(), k -> {
			try {
				return createIfAbsent(input);
			} catch (Exception e) {
				// What to do after catching
			}
			return input;
		});
	}
```

vinayakapawar replied:
1. I believe getCpFromDb will throw an unchecked exception (OpenSpecimenException).

2. If #1 is true, I don't see any value in catching the exception when you don't know what to do. If you don't catch, the exception gets propagated to the caller and its caller until someone handles it or it falls of the method call chain.

3. You are getting stumped for a wide ball.

vinayakapawar commented on src/main/java/com/krishagni/openspecimen/msk2/importer/CarsImporter.java:
```diff
@@ -0,0 +1,278 @@
+	//
+	/////////////////////////
+	
+	private void checkSr(CollectionProtocolEventDetail event) throws Exception {
```
1. The input to this method should be SpecimenRequirementDetail.
2. Method name should be getSr(SpecimenRequirementDetail input)

vinayakapawar commented on src/main/java/com/krishagni/openspecimen/msk2/importer/CarsImporter.java:
```diff
@@ -0,0 +1,278 @@
+	
+	private void checkSr(CollectionProtocolEventDetail event) throws Exception {
+		SpecimenRequirementDetail srDetail = srMap.get(event.getSpecimenRequirements().iterator().next().getName());
+		
```
If you follow above comment, this circus need not be done.

vinayakapawar commented on src/main/java/com/krishagni/openspecimen/msk2/importer/CarsImporter.java:
```diff
@@ -0,0 +1,278 @@
+				.getCollectionProtocolDao()
+				.getCpeByShortTitleAndEventLabel(event.getCpShortTitle(), event.getEventLabel());
+		
+		Set<SpecimenRequirement> specimenRequirements = cpe != null ? cpe.getSpecimenRequirements() : Collections.emptySet();
```
Get only top-level requirements. You will getting information about only the primary specimens.

vinayakapawar commented on src/main/java/com/krishagni/openspecimen/msk2/importer/CarsImporter.java:
```diff
@@ -0,0 +1,278 @@
+				return req;
+			}
+		}
+		
```
Where possible write concise and readable code by using stream APIs

vinayakapawar commented on src/main/java/com/krishagni/openspecimen/msk2/importer/CarsImporter.java:
```diff
@@ -0,0 +1,278 @@
+		return null;
+	}
+	
+	private SpecimenRequirementDetail createSr(List<SpecimenRequirementDetail> sprDetails) throws Exception {
```
Input should be just SR detail object and not the collection.

vinayakapawar commented on src/main/java/com/krishagni/openspecimen/msk2/importer/CarsImporter.java:
```diff
@@ -0,0 +1,278 @@
+		return resp.getPayload();
+	}
+
+	private SpecimenRequirementDetail updateSr(List<SpecimenRequirementDetail> srDetails, SpecimenRequirementDetail srFromDb) throws Exception {
```
The order of parameters to updateSr should be updateSr(existingSr, newSr)

vinayakapawar commented on src/main/java/com/krishagni/openspecimen/msk2/importer/CarsImporter.java:
```diff
@@ -0,0 +1,278 @@
+			CollectionProtocolEventDetail event) {
+		
+	}
+	
```
Same comments apply to events as given for SR and CP.

vinayakapawar commented on src/main/java/com/krishagni/openspecimen/msk2/importer/CarsImporter.java:
```diff
@@ -0,0 +1,278 @@
+	private final static String SPECIMEN_STORAGE_TYPE_VIRTUAL = "Virtual";
+	
+	private final static String NOT_SPECIFIED = "Not Specified";
+}
```
The order of methods in the file should be in the same order that they are first used and grouped together.

So this means, you will have 3 blocks

// Block 1: CP related private methods

// Block 2: CPE related private methods

// Block 3: SR related private methods.

vinayakapawar commented on build.gradle:
```diff
@@ -0,0 +1,128 @@
+  compile group: 'org.springframework', name: 'spring-web', version:springVersion
+  compile group: 'org.springframework', name: 'spring-webmvc', version:springVersion
+  compile group: 'org.springframework', name: 'spring-orm', version:springVersion
+  compile group: 'org.springframework.security', name: 'spring-security-core', version: '3.2.7.RELEASE'
```
Remove dependencies that you really don't require in this plugin.

Swapnil-ingle committed:
SI: Made the code changes commented in pull request #1
1. Rather than overloading event using three functions: toCp(), toSr() and toEvent().
2. Removed the variables that were pre-initialised at the starting of the block.
3. The order of the field initialisation changed to cp short title -> event label -> sr name -> type.
4. Changed from checkCp(CollectionProtocolEventDetail event, CarsDetail object) to getCp(CollectionProtocolDetail cp)
5. Instead of having several nested Ifs statements, added the map.computeIfAbsent() functionality.
6. Getting only top-level requirements from CPE.
7. Rewritten the code in getSrFromDb() using stream APIs.
8. Changed from checkSr() -> getSr(SpecimenRequirementDetail input)
9. Changed the createSr() from having Collections as a param to having SpecimenRequirementDetail as param
10. Changed updateSr() method to updateSr(existingSr, newSr)
11. Changed updateEvent() method to updateEvent(existingEvent, newEvent)
12. Removed unnecessary dependancies from build.gradle
13. The code is now in three specific blocks CP, CPE and SR related method respectively.

Swapnil-ingle committed:
SI: Refactored the CARS code to read from DB instead of CSV
1. Added a CarsReader class that handles reading from DB
2. Removed msk-cars.xml
3. Removed messages.properties

Swapnil-ingle committed:
SI: Added the datetime conditions in MS-SQL DB and retrieved it into detail object.
1. TIMEPOINT_CR_DATE
2. PROCEDURE_CR_DATE
3. TIMEPOINT_UPDATE
4. PROCEDURE_UPDATE

Swapnil-ingle committed:
SI: The database credentials are now taken from UI instead of hardcoded strings.

Swapnil-ingle committed:
SI: Changed the code structure to calculate Events or SR (Added/Updated/Deleted) and then apply those changes.
